2 Bugfixes for ContextImpl and RootImpl

mpetris opened the pull request:
Hi, 
this Pull Request suggests two bugfixes: 

ContextImpl: in the case of a LeafNode calling childNodes throws an Exception, therefore we check childNodeSize first

RootImpl: access to index 2 fails for elements with no children and no attributes because array size is 2 then containing only tagname and element id, therefore we check for length before trying access at idx 2

mpetris committed:
bugfix: in the case of a LeafNode calling childNodes throws an Exception, therefore check childNodeSize first

mpetris committed:
access to index 2 fails for elements with no children and no attributes because array size is 2 then conaining only tagname and element id
therefore we check for length before trying access at idx 2

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/vaadin/serverside-elements?pullRequest=22) <br/>All committers have signed the CLA.

Legioth commented:
Hi. Thank you for making these fixes. It would be really awesome if you could also add unit tests for these cases. I can also do that even though I won't have time to get that done immediately.

mpetris committed:
added tests to check changes for parsing DOM leafs without attributes and adding LeafNode work
in order to test LeafNode an upgrade of the referenced vaadin version because of the newer jsoup dependency was necessary

mpetris commented:
Hi @Legioth, 
ok I added a test for each fix. 
In order to make the second test addLeafToRootNode actually relevant I needed to upgrade the referenced Vaadin version to 8.6.2 in the pom files. This is because the addressed problem only occurrs with the newer jsoup version 1.11.2 referenced by Vaadin 8.6.2 and not with the older 1.8.1 referenced by Vaadin 8.0.5.

Legioth commented on elements-demo/pom.xml:
```diff
@@ -5,12 +5,12 @@
 	<properties>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-		<vaadin.version>8.0.5</vaadin.version>
+		<vaadin.version>8.6.2</vaadin.version>
```
```suggestion
 		<vaadin.version>8.0.5</vaadin.version>
```

Legioth commented on elements/pom.xml:
```diff
@@ -10,7 +10,7 @@
 	<properties>
 		<project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
-		<vaadin.version>8.0.5</vaadin.version>
+		<vaadin.version>8.6.2</vaadin.version>
```
```suggestion
 		<vaadin.version>8.0.5</vaadin.version>
```

Legioth commented on elements/src/main/java/org/vaadin/elements/impl/Context.java:
```diff
@@ -37,8 +37,9 @@ public void wrapChildren(NodeImpl node) {
             Node soupChild = queue.poll();
-
-            queue.addAll(soupChild.childNodes());
+            if (soupChild.childNodeSize() > 0) {
```
```suggestion
             // Only relevant for newer JSoup versions
             if (soupChild.childNodeSize() > 0) {
```

Legioth commented on elements/src/test/java/org/vaadin/elements/RootTest.java:
```diff
@@ -70,4 +72,20 @@ public void innerHtmlElementOrder() {
+
+    @Test
+    public void addLeafToRootNode() {
+        root.setInnerHtml("<span>Text-Leaf</span>");
```
```suggestion
         // Only relevant for newer JSoup versions
         root.setInnerHtml("<span>Text-Leaf</span>");
```

Legioth reviewed (changes requested):
I think it would be best to still keep the old version numbers, but add notes in the code for the things that are only relevant with newer JSoup versions.

I've added suggestions for the relevant changes.

Legioth committed:
Update elements-demo/pom.xml

Co-Authored-By: mpetris <marco.petris@web.de>

Legioth committed:
Update elements/pom.xml

Co-Authored-By: mpetris <marco.petris@web.de>

Legioth committed:
Update elements/src/main/java/org/vaadin/elements/impl/Context.java

Co-Authored-By: mpetris <marco.petris@web.de>

Legioth committed:
Update elements/src/test/java/org/vaadin/elements/RootTest.java

Co-Authored-By: mpetris <marco.petris@web.de>

Legioth commented on elements/pom.xml:
```diff
@@ -74,7 +74,7 @@
 			<groupId>org.jsoup</groupId>
 			<artifactId>jsoup</artifactId>
-			<version>1.8.1</version>
+			<version>1.11.2</version>
```
Should maybe not define any JSoup version at all here, since any explicitly chosen version might be incompatible with what the used Vaadin 8 version uses. 

I don't think an explicit dependency is needed here since JSoup should come in as a transitive dependency from `vaadin-server`?

mpetris replied:
yes, that's right, the dependency comes already with vaadin-server

Legioth commented:
Thanks again for contributing these fixes!

Legioth commented:
Released as https://vaadin.com/directory/component/elements-add-on/0.2.1
